### PR TITLE
Refactoring & adding SPARSE_BINDING to QueueType

### PR DESCRIPTION
**Pull request by Neo-Zhixing**
We now have sparse binding thanks to #3521 but currently 'SPARSE_BINDING' is missing from 'QueueType'. This PR adds the 'SPARSE_BINDING' bit to QueueType by converting it from a enum to a bitflags struct.

Reference: https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkQueueFlagBits.html

PR checklist:
- [x] `make` succeeds (on *nix)
- [ ] `make reftests` succeeds
- [x] tested examples with the following backends: Metal, DX12, DX11, Vulkan

### CODE REVIEW

File: src/backend/empty/src/lib.rs [hunk 1]
```diff
@@ -575,7 +575,7 @@ impl device::Device<Backend> for Device {
 pub struct QueueFamily;
 impl queue::QueueFamily for QueueFamily {
     fn queue_type(&self) -> queue::QueueType {
```
Comment (kvark) [inline src/backend/empty/src/lib.rs:577]: this type and function should no longer be called "queue type". It's a set of flags now, so `QueueFlags`?

File: src/hal/src/queue/mod.rs [hunk 2]
```diff
@@ -19,38 +19,40 @@ use std::{any::Any, fmt};
...
+        const GRAPHICS = 0b1;
+        /// Queue supports compute operations
+        const COMPUTE = 0b10;
+        /// Queue supports transfer operations
+        const TRANSFER = 0b100;
+        /// Queue supports sparse resource memory management operations
+        const SPARSE_BINDING = 0b1000;
+
+        /// Queue supports both graphics operations and transfer operations
+        const GRAPHICS_TRANSFER = Self::GRAPHICS.bits | Self::TRANSFER.bits;
```
Comment (kvark) [inline src/hal/src/queue/mod.rs:36]: This shouldn't be needed. All graphics queues are required to support transfers.

File: src/hal/src/queue/mod.rs [hunk 3]
```diff
@@ -19,38 +19,40 @@ use std::{any::Any, fmt};
...
+
+        /// Queue supports both compute operations and transfer operations
+        const COMPUTE_TRANSFER = Self::GRAPHICS.bits | Self::COMPUTE.bits;
```
Comment (kvark) [inline src/hal/src/queue/mod.rs:39]: same here, all compute queues are required to support transfers

**Review by kvark — COMMENTED** [review]
Thank you for contributing!
Do we really need to make these to be flags? Alternatively, we could make `supports_sparse_binding` function to `trait QueueFamily`. That would be less intrusive.

File: src/backend/empty/src/lib.rs (hunk 1 above)
Comment (Neo-Zhixing) [reply src/backend/empty/src/lib.rs]: name changed

**Comment by Neo-Zhixing** [comment]
I agree that simply adding a `supports_sparse_binding` function could be easier. However this is a better match to the original Vulkan APIs. I'm not sure which one has a higher priority - API stability or matching Vulkan APIs?

File: src/backend/dx11/src/lib.rs [hunk 4]
```diff
@@ -1135,8 +1135,8 @@ impl window::PresentationSurface<Backend> for Surface {
 pub struct QueueFamily;
 
 impl queue::QueueFamily for QueueFamily {
-    fn queue_type(&self) -> queue::QueueType {
-        queue::QueueType::General
+    fn queue_flags(&self) -> queue::QueueFlags {
```
Comment (kvark) [inline src/backend/dx11/src/lib.rs:1138]: maybe just `fn flags(&self)`?

**Review by kvark — COMMENTED** [review]
This is a tough choice, but also a relatively small piece of the API.
I'm good with proceeding. Let's just fix the CI (failures on Windows) and maybe rename `queue_flags` to just `flags()`?

**Comment by Neo-Zhixing** [comment]
I've opened another PR (#3667) that simply added `supports_sparse_binding` function to `trait QueueFamily` as suggested. Please merge that PR and close this one if preferred.

**Comment by kvark** [comment]
Thank you! This is very nice of you to provide the alternative. I'm not feeling strongly about this decision, but I merged #3667 as less of an intrusive solution.
Philosophically, we wanted to follow Vulkan were possible, so that we can get Vulkan Portability nicely, and it's easy to find answers about the guarantees by reading the Vulkan spec. In practice, both of these needs dim compared to the need to have `wgpu` to run efficiently on gfx, so some parts of gfx get slightly adjusted to make this path fast and efficient (like the swapchain API). So it's fine to have the queue types stay.

**Pull request closed without merging**